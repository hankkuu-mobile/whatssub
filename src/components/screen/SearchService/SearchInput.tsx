import * as React from 'react';

import { IC_MAGNIFIER } from '../../../utils/Icons';
import styled from 'styled-components/native';

const Container = styled.View`
  background-color: ${({ theme }) => (theme.debug ? '#ddd' : '#f3f5f7')};
  height: 42;
  margin-left: 20;
  margin-right: 20;
  border-radius: 8;

  flex-direction: row;
`;

const MagContainer = styled.View`
  background-color: ${({ theme }) => (theme.debug ? 'red' : 'transparent')};
  height: 24;
  width: 24;
  margin-left: 8;
  margin-right: 2;
  margin-top: 9;
  margin-bottom: 9;
  justify-content: center;
  align-items: center;
`;

const Magnifier = styled.Image`
  background-color: ${({ theme }) => (theme.debug ? 'green' : 'transparent')};
  width: 16;
  height: 16;
  margin-bottom: 2;
`;

const Input = styled.TextInput`
  background-color: ${({ theme }) => (theme.debug ? '#aaa' : 'transparent')};
  flex-grow: 1;
  align-self: center;
  font-size: 16;
`;

const ResetContainer = styled.View`
  background-color: ${({ theme }) => (theme.debug ? 'blue' : 'transparent')};
  height: 24;
  width: 24;
  margin-left: 2;
  margin-right: 11;
  margin-top: 9;
  margin-bottom: 9;
  justify-content: center;
  align-items: center;
`;

const Reset = styled.TouchableOpacity`
  background-color: ${({ theme }) => (theme.debug ? 'yellow' : '#c6ccd1')};
  border-radius: 20;
  width: 20;
  height: 20;
  justify-content: center;
  align-items: center;
`;

const ResetText = styled.Text`
  color: white;
`;

interface SearchInputProps {
  value: string;
  onDebounceOrOnReset: (value: string) => void;
  style?: any;
}

// reference : https://dev.to/gabe_ragland/debouncing-with-react-hooks-jci
function useDebounce(value, delay) {
  const [debouncedValue, setDebouncedValue] = React.useState(value);

  React.useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);
    return () => {
      clearTimeout(handler);
    };
  }, [value]);

  return debouncedValue;
}

function SearchInput(props: SearchInputProps) {
  const [value, setValue] = React.useState(props.value);
  const debouncedValue = useDebounce(value, 400);

  React.useEffect(() => {
    props.onDebounceOrOnReset(debouncedValue);
  }, [debouncedValue]);

  React.useEffect(() => {
    setValue(props.value);
  }, [props.value]);

  return (
    <Container style={props.style}>
      <MagContainer>
        <Magnifier source={IC_MAGNIFIER} />
      </MagContainer>
      {/* TODO: add placeholder string to Sting */}
      <Input
        testID={'SEARCH_INPUT'}
        value={value}
        onChangeText={(text) => {
          setValue(text);
        }}
        placeholder={'검색어를 입력해주세요.'}
        placeholderTextColor={'#cdd2d7'}
      />
      {props.value !== '' && (
        <ResetContainer>
          <Reset
            testID='RESET_BUTTON'
            onPress={() => {
              props.onDebounceOrOnReset('');
            }}
          >
            <ResetText>X</ResetText>
          </Reset>
        </ResetContainer>
      )}
    </Container>
  );
}

export default SearchInput;
