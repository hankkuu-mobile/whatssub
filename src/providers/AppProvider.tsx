import * as Font from 'expo-font';

import React, { useEffect, useReducer, useState } from 'react';

import { AsyncStorage } from 'react-native';
import { ThemeType } from '../types';

interface Context {
  state: State;
  changeTheme: (theme?: ThemeType) => void;
}
const AppContext = React.createContext<Context>(null);
const AppConsumer = AppContext.Consumer;

interface ChangeThemeTypeAction {
  type: 'change-theme-type-mode';
  payload: {
    theme?: ThemeType;
  };
}

interface Props {
  theme?: ThemeType;
  children?: any;
  doNotWaitFont?: boolean;
}

interface State {
  theme?: ThemeType;
}

export const initialState: State = {
  theme: ThemeType.LIGHT,
};

const reducer = (state: State, action: ChangeThemeTypeAction) => {
  switch (action.type) {
    case 'change-theme-type-mode':
      // if payload.theme doesn't exist, toggle
      if (!action.payload.theme) {
        return {
          ...state,
          theme:
            state.theme === ThemeType.LIGHT
              ? ThemeType.DARK
              : ThemeType.LIGHT,
        };
      }
      // else change it with payload.theme
      return {
        ...state,
        theme: action.payload.theme,
      };
    default:
      return state;
  }
};

function AppProvider(props: Props) {
  const [state, dispatch] = useReducer(reducer, initialState);
  const [fontLoading, setFontLoading] = useState(true);

  // by encapsulating dispatch, it is easier to use (such as passing through screenProps)
  const changeTheme: (theme?: ThemeType) => void = (theme) => {
    dispatch({
      type: 'change-theme-type-mode',
      payload: {
        theme,
      },
    });
  };

  // initialize theme
  useEffect(() => {
    changeTheme(props.theme ? props.theme : initialState.theme);
  }, [props.theme]);

  // Load font and then use saved theme from local storage
  const getCurrentThemeType = async () => {
    const value = (await AsyncStorage.getItem('theme')) as ThemeType;
    value && changeTheme(value);
  };
  const loadFonts = async () => {
    await Font.loadAsync({
      'spoqa-han-sans-bold': require('../../assets/fonts/SpoqaHanSans/SpoqaHanSans-Bold.ttf'),
      'spoqa-han-sans-regular': require('../../assets/fonts/SpoqaHanSans/SpoqaHanSans-Regular.ttf'),
    });

    setFontLoading(false);
  };
  useEffect(() => {
    loadFonts();
    getCurrentThemeType();
  }, []);

  useEffect(() => {
    AsyncStorage.setItem('theme', state.theme);
  }, [state.theme]);

  return (
    <AppContext.Provider
      value={{
        state,
        changeTheme,
      }}
    >
      {fontLoading && !props.doNotWaitFont ? null : props.children}
    </AppContext.Provider>
  );
}

export { AppConsumer, AppProvider, AppContext };
