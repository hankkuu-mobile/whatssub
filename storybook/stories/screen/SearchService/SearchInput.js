import React, { useContext, useState } from 'react';

import { Button } from 'dooboo-native-widgets';
import DebugButton from '../../../../src/components/shared/DebugButton';
import SearchInput from '../../../../src/components/screen/SearchService/SearchInput';
import { storiesOf } from '@storybook/react-native';
import styled from 'styled-components/native';

const Container = styled.SafeAreaView`
  flex: 1;
`;

const SearchInputWithState = () => {
  const [value, setValue] = useState('');
  return <SearchInput value={value} onDebounceOrOnReset={setValue} />;
};

const ContainerDeco = (storyFn) => <Container>{storyFn()}</Container>;

storiesOf('SearchService', module)
  .addDecorator(ContainerDeco)
  .add('â””SearchInput', () => (
    <>
      <SearchInputWithState />
      <DebugButton />
    </>
  ));
