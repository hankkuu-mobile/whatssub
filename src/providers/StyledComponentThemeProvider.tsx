import React, { useContext, useState } from 'react';

import { AppContext } from './AppProvider';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import { createTheme } from '../theme';

const StyledComponentThemeProvider = (props) => {
  const { state } = useContext(AppContext);
  const [debug, setDebug] = useState(false);
  return (
    <ThemeProvider theme={{ ...createTheme(state.theme), debug, setDebug }}>
      {props.children}
    </ThemeProvider>
  );
};

StyledComponentThemeProvider.propTypes = {
  children: PropTypes.node,
};

export default StyledComponentThemeProvider;
