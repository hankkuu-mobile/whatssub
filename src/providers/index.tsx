import { AppConsumer, AppContext, AppProvider } from './AppProvider';

import PropTypes from 'prop-types';
import React from 'react';
import StyledComponentThemeProvider from './StyledComponentThemeProvider';
import { ThemeType } from '../types';

interface Props {
  theme?: ThemeType;
  children?: any;
  doNotWaitFont?: boolean;
}

const InitialProviders = ({ children, ...rest }: Props) => (
  <AppProvider {...rest}>
    <StyledComponentThemeProvider>{children}</StyledComponentThemeProvider>
  </AppProvider>
);

InitialProviders.propTypes = {
  children: PropTypes.node,
};

export { AppProvider, AppConsumer, AppContext, InitialProviders };
