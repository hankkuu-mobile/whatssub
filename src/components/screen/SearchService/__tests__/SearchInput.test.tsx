import * as React from 'react';

import { Text, View } from 'react-native';
import { fireEvent, render, wait } from '@testing-library/react-native';
import renderer, { act } from 'react-test-renderer';

import { InitialProviders } from '../../../../providers';
import SearchInput from '../SearchInput';
import i18n from 'i18n-js';

i18n.locale = 'ko';
const createTestProps = ({ value = '' }: { value: string }) => ({
  onDebounceOrOnReset: jest.fn(),
  value: value,
});

const component = (props) => (
  <InitialProviders doNotWaitFont>
    <SearchInput {...props} />
  </InitialProviders>
);

describe('[SearchInput] render', () => {
  it('component and snapshot matches', () => {
    const rendered: renderer.ReactTestRendererJSON = renderer
      .create(component(createTestProps({ value: '' })))
      .toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });
});

describe('[SearchInput] interaction', () => {
  it('should show reset button when input `value` exists', () => {
    const rendered = render(
      component(createTestProps({ value: 'initial value' })),
    );
    expect(rendered.getByTestId('RESET_BUTTON')).toBeTruthy();
  });

  it('should not show reset button when empty input value', () => {
    const rendered = render(component(createTestProps({ value: '' })));
    expect(rendered.queryByTestId('RESET_BUTTON')).toBeNull();
  });
});
