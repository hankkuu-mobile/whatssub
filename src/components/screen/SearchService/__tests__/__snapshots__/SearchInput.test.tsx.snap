// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SearchInput] render component and snapshot matches 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#f3f5f7",
        "borderRadius": 8,
        "flexDirection": "row",
        "height": 42,
        "marginLeft": 20,
        "marginRight": 20,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "height": 24,
          "justifyContent": "center",
          "marginBottom": 9,
          "marginLeft": 8,
          "marginRight": 2,
          "marginTop": 9,
          "width": 24,
        },
      ]
    }
  >
    <Image
      source={1}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "height": 16,
            "marginBottom": 2,
            "width": 16,
          },
        ]
      }
    />
  </View>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="검색어를 입력해주세요."
    placeholderTextColor="#cdd2d7"
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "alignSelf": "center",
          "backgroundColor": "transparent",
          "flexGrow": 1,
          "fontSize": 16,
        },
      ]
    }
    testID="SEARCH_INPUT"
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;
